src/language-js/postprocess.js

```diff
@@ -5,8 +5,42 @@ const {
   getNextNonSpaceNonCommentCharacter
 } = require("../common/util");
 const { composeLoc, locEnd } = require("./loc");
+const { isTypeCastComment } = require("./comments");
 
 function postprocess(ast, options) {
+  // Keep Babel's non-standard ParenthesizedExpression nodes only if they have Closure-style type cast comments.
+  if (options.parser !== "typescript" && options.parser !== "flow") {
+    const startOffsetsOfTypeCastedNodes = new Set();
+
+    // Comments might be attached not directly to ParenthesizedExpression but to its ancestor.
+    // E.g.: /** @type {Foo} */ (foo).bar();
+    // Let's use the fact that those ancestors and ParenthesizedExpression have the same start offset.
+
+    visitNode(ast, node => {
+      if (
+        node.leadingComments &&
+        node.leadingComments.some(isTypeCastComment)
+      ) {
+        startOffsetsOfTypeCastedNodes.add(node.start);
+      }
+    });
+
+    visitNode(ast, node => {
```

Comment:
We could probably do this in a single walk, since the ancestor will have been visited first.

Reply:
No, it won't.

Reply:
No, it won't.